File: src/main/java/Utilities/Preprocessor.java
package main.java.Utilities;
/*
 * Copyright (C) 2013 Rochester Institute of Technology
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * See the GNU General Public License for more details.
 */

import java.awt.Color;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Hashtable;
import java.util.Map;
import java.util.Random;
import java.util.Vector;

import main.java.Electro1D.Protein;
import main.java.Electro2D.Electro2D;

/**
 * This class generates the preprocessed proteome files created after a genBank,
 * fasta, or protein database file has been run. In doing this the time to load
 * a proteome (if already loaded once) will be cut down. (Hopefully...)
 * 
 * BH Note: In addition, the class serves as a temporary container for several
 * data vectors that are used in Electro1D and Electro2D and holds the
 * calculations for molecular weight and pI.
 * 
 * 
 *
 * @author Jill Zapoticznyj
 */

public class Preprocessor {

	public static void init() {
		// just initialize this class to speed later class loading
	}
	
	private static Random rand = new Random();

	private final static double[] partialCharges = new double[256];

	private final static double[] mws = new double[256];

	static {
		mws['A'] = 71.0938;
		mws['C'] = 103.1538;
		mws['D'] = 115.1036;
		mws['E'] = 129.1304;
		mws['F'] = 147.1914;
		mws['G'] = 57.067;
		mws['H'] = 137.156;
		mws['I'] = 113.1742;
		mws['K'] = 128.1888;
		mws['L'] = 113.1742;
		mws['M'] = 131.2074;
		mws['N'] = 114.1188;
		mws['P'] = 97.1316;
		mws['Q'] = 128.1456;
		mws['R'] = 156.2022;
		mws['S'] = 87.0932;
		mws['T'] = 101.12;
		mws['V'] = 99.1474;
		mws['W'] = 186.228;
		mws['Y'] = 163.1908;
	}
	private final static Color[] colors = { Color.black, Color.blue, Color.cyan, Color.green, Color.magenta,
			Color.orange, Color.pink, Color.red };

	private static Color randomColor() {
		return colors[rand.nextInt(colors.length)];
	}

	private static final String EXTENTION = ".e2d";
	private static final int HEADER_WIDTH = 12;
	private static final int LINE_LENGTH = 50;

	private final String FILE_HEADER        = "FILE:       ";
//    private final String NUMENZYME_HEADER = "NUMENZYMES: ";
	private final String PROTTITLE_HEADER   = "NAME:       ";
	private final String FUNCTION_HEADER    = "FUNCTION:   ";
	private final String SEQUENCE_HEADER    = "SEQUENCE:   ";
	private final String MOLWT_HEADER       = "MOLWEIGHT:  ";
	private final String PIVAL_HEADER       = "PIVAL:      ";
	private final String SEQ_SEPARATOR     = "-----";

	Vector<String> sequences;
	Vector<String> sequenceTitles;
	Vector<String> piValues;
	Vector<String> molecularWeights;
	Vector<String> functions;
	Vector<String> concentrations;

	private String inputName;
	private double[] minmax;
	private boolean is1D;

	private String line;

	private boolean isE2D;

	private String e2dOutFileName;

	Preprocessor(File inputFile, int fileNum) {
		this(null, inputFile, fileNum <= 0);
	}

	/**
	 * 
	 * @param e         an Electro2D object or null if Electro1D
	 * @param inputFile
	 * @param fileNum   0 (Electro1D) or 1 or 2
	 */
	Preprocessor(Electro2D e, File inputFile, boolean is1D) {
		this.is1D = is1D;
		sequences = (e == null ? new Vector<>() : e.getSequences());
		sequenceTitles = (e == null ? new Vector<>() : e.getSequenceTitles());
		molecularWeights = (e == null ? new Vector<>() : e.getMolecularWeights());
		piValues = (is1D ? null : e == null ? new Vector<>() : e.getPiValues());
		functions = (e == null ? new Vector<>() : e.getFunctions());
		inputName = (e == null ? inputFile.getName() : e.getLastFileLoaded());
		concentrations = (e == null ? new Vector<>() : null);
	}

	public void writeToE2DFile(String outFileName) {
		if (outFileName == null) {
			outFileName = "data" + File.separator + inputName.replace('.', '_') + EXTENTION;
		}
		File theFile = new File(outFileName);
		System.out.println("Preprocessor.writeToFile creating " + theFile);
		if (theFile.exists())
			return;
		try (PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter(theFile)), true)) {
			out.println(FILE_HEADER + inputName);
			out.println(SEQ_SEPARATOR);
			for (int i = 0, n = sequences.size(); i < n; i++) {
				out.println(PROTTITLE_HEADER + sequenceTitles.get(i));
				writeRecord(out, FUNCTION_HEADER, functions.get(i));
				String seq = sequences.get(i);
				writeRecord(out, SEQUENCE_HEADER, seq);
				out.println(MOLWT_HEADER + molecularWeights.get(i));
				out.println(PIVAL_HEADER + (piValues == null ? getPI(seq) : piValues.get(i)));
				out.println(SEQ_SEPARATOR);
			}
		} catch (IOException e) {
			System.err.println("Preprocessor.wrteToE2DFile Error writing to file " + e.getMessage());
			return;
		}
	}

	/**
	 * write the header/data record
	 * @param out
	 * @param header
	 * @param data
	 */
	private static void writeRecord(PrintWriter out, String header, String data) {
		int length = data.length();
		int pt = 0;
		while (length > 0) {
			int n = Math.min(length, LINE_LENGTH);
			out.println(header + data.substring(pt, pt = pt + n));
			length -= n;
		}
	}

	public int readFromFile(BufferedReader reader, Vector<String> sequencesOut, Vector<Protein> proteins, Electro2D electro2D, int fileNum) {
		String err = null;
		isE2D = true;
		StringBuffer buf = new StringBuffer();
		try {
			while (true) {
				if ((line = reader.readLine()).indexOf(FILE_HEADER) != 0) {
					err = "Preprocessor.readFromeFile Invalid file format.  Missing FILE: line";
					break;
				}
				if (electro2D != null)
					electro2D.setLastFileLoaded(line.substring(HEADER_WIDTH));
				while ((line = reader.readLine()) != null) {
					if (line.startsWith("-"))
						continue;
					sequenceTitles.add(readE2D(reader, PROTTITLE_HEADER, buf));
					functions.add(readE2D(reader, FUNCTION_HEADER, buf));
					sequences.add(readE2D(reader, SEQUENCE_HEADER, buf));
					molecularWeights.add(readE2D(reader, MOLWT_HEADER, buf));
					if (is1D)
						reader.readLine(); // skip piValue
					else
						piValues.add(readE2D(reader, PIVAL_HEADER, buf));
				}
				finalizeRead(sequencesOut, proteins, electro2D, fileNum);
				// electro2D.setLastFileLoaded(
				reader.close();
				break;
			}
		} catch (Exception e) {
			e.printStackTrace();
			err = "Error reading from file.  Be sure you entered the file name correctly. " + e.getMessage();
		}
		if (err != null)
			System.out.println(err);
		return sequences.size();
	}

	/**
	 * Read an E2D record, consisting of some number of lines starting with the same
	 * header, as in PDB files.
	 * 
	 * @param reader
	 * @param key    "NAME:" for example
	 * @param buf
	 * @return
	 * @throws IOException
	 */
	private String readE2D(BufferedReader reader, String key, StringBuffer buf) throws IOException {
		buf.setLength(0);
		while (line.startsWith(key)) {
			buf.append(line.substring(HEADER_WIDTH));
			line = reader.readLine();
		}
		return buf.toString();
	}

	private void getPIandMW() {
		minmax = (is1D ? null : new double[] { -1, Double.MAX_VALUE, -1, Double.MAX_VALUE });
		for (int i = 0; i < sequences.size(); i++) {
			String mw, pI = "";
			if (isE2D) {
				mw = molecularWeights.get(i);
				if (!is1D)
					pI = piValues.get(i);
			} else {
				String temp = sequences.get(i);
				mw = getMW(temp);
				molecularWeights.add(mw);
				if (!is1D) {
					pI = getPI(temp);
					piValues.add(pI);
				}
			}
			if (minmax != null) {
				double d = Double.parseDouble(mw);
				if (d <= minmax[0]) {
					minmax[0] = d;
				}
				if (d >= minmax[1]) {
					minmax[1] = d;
				}
				d = Double.parseDouble(pI);
				if (d <= minmax[2]) {
					minmax[2] = d;
				}
				if (d >= minmax[3]) {
					minmax[3] = d;
				}
			}
		}
	}
	
	int finalizeRead(Vector<String> sequencesOut, Vector<Protein> proteins, Electro2D electro2D, int fileNum) {
		getPIandMW();
		int nSeq = sequences.size();
		switch (fileNum) {
		case GenomeFileParser.SEQUENCES_ONLY:
			// Mass Spec
			sequencesOut.addAll(sequences);
			return nSeq;
		case GenomeFileParser.PROTEINS_ONLY:
			// Electro1D
			for (int i = 0; i < nSeq; i++) {
				proteins.add(new Protein(sequenceTitles.get(i), "", "",
						(int) Double.parseDouble(molecularWeights.get(i)), randomColor()));
				if (concentrations.size() > 0)
					proteins.get(i).setConcentration(Integer.parseInt(concentrations.get(i)));
			}
			return nSeq;
		case GenomeFileParser.ELECTRO2D_FILE_1:
			electro2D.setSequences(sequences);
			electro2D.setSequenceTitles(sequenceTitles);
			electro2D.setMolecularWeights(molecularWeights);
			electro2D.setPiValues(piValues);
			electro2D.setFunctionValues(functions);
			break;
		case GenomeFileParser.ELECTRO2D_FILE_2:
			electro2D.setSequences2(sequences);
			electro2D.setSequenceTitles2(sequenceTitles);
			electro2D.setMolecularWeights2(molecularWeights);
			electro2D.setPiValues2(piValues);
			electro2D.setFunctionValues2(functions);
			break;
		}
		electro2D.setLastFileLoaded(inputName);
		electro2D.setMaxAndMinVals(minmax[1], minmax[0], minmax[3], minmax[2]);
		if (!isE2D && !JSUtil.isJS)
			writeToE2DFile(e2dOutFileName);
		return nSeq;
	}

	private final static Map<String, String> htPIcache = new Hashtable<>();
	private final static Map<String, String> htMWcache = new Hashtable<>();

	/**
	 * This method calculates the molecular weight from inputed sequence
	 *
	 * @param seq protein sequence
	 *
	 * @return returns the molecular weight
	 */
	public static String getMW(String seq) {
		String wt = htMWcache.get(seq);
		if (wt != null)
			return wt;
		int len = seq.length();

		// Molecular weight of protein, starting with 1 unit of H2O
		
		double mW = 18;

//     Determine the molecular weight for each AA until reach
//     end of sequence. Add the weight for each AA to the value
//	   of total weight. Since a water molecule (MW 18) is lost with
//	   each bond, the weight given for each AA is its molecular
//	   weight minus that of water. The MW of one water molecule is
//	   then added to the total molecular weight of the protein sequence.
//	   If the AA character does not match one of the 20 accepted
//	   abbreviations, then a weight of 0 is given for that AA.

		for (int f = 0; f < len; f++) {
			mW += mws[seq.charAt(f)];
		}

		String mWstring = roundOff(mW, 2);
		htMWcache.put(seq, mWstring);
		return mWstring;
	}
	
	private static void calculatePartialCharges(double pH) {

		// pK values were obtained from Bjellqvist, B., Basse, B., Olsen, E.,
		// Celis, J., Reference points for comparisons of two-dimensional
		// maps of proteins from different human cell types defined in a
		// pH scale where isoelectric points correlate with polypeptide
		// compositions, Electro1DMain 1994, 15, 529-539.
		

		partialCharges['D'] = getPartialCharge(false, 4.05, pH); // ASP
		partialCharges['C'] = getPartialCharge(false, 9.00, pH); // CYS
		partialCharges['E'] = getPartialCharge(false, 4.75, pH); // GLU
		partialCharges['Y'] = getPartialCharge(false, 10.0, pH); // TYR
		
		partialCharges['R'] = getPartialCharge(true, 12.0, pH); // ARG
		partialCharges['H'] = getPartialCharge(true, 5.98, pH); // His
		partialCharges['K'] = getPartialCharge(true, 10.0, pH); // LYS

		partialCharges['\1'] = getPartialCharge(true, 8.2, pH); // N terminus, NH3+
		partialCharges['\2'] = getPartialCharge(false, 3.2, pH); // C terminus, CO2-
	}

	/**
	 * @param isBasic -- that is, "is a basic amino acid," meaning that it is protonated, like lysine.
	 * @param pK
	 * @param pH
	 * @return
	 */
	private static double getPartialCharge(boolean isBasic, double pK, double pH) { 
		return 1 / (isBasic ? (1 + Math.pow(10, pH - pK)) : -(1 + Math.pow(10, pK - pH))); 
	}

	/**
	 * This method calculates the pI from inputed sequence
	 *
	 * @param seq protein sequence
	 * @return returns the pI value
	 */
	public static String getPI(String seq) {

		String pi = htPIcache.get(seq);
		if (pi != null)
			return pi;

		double lowpH = 0, pH = 7, highpH = 14;

		int plen = seq.length();

		double charge = 0;

		// Calculate total charge at varying pH values until the
		// charge is within 0.005 of 0
		while (true) {
			// Calculate the charge for each AA until reach end of sequence
			// Add the charge for each AA to the value of total charge
			calculatePartialCharges(pH);
			charge = 0;
			for (int a = plen; --a >= 0;) {
				// Determine appropriate pK value for current AA
				// If AA not acid or base, set to neutral
				// and give default pK of 0
				charge += partialCharges[seq.charAt(a)];
			}
			charge += partialCharges['\1'] + partialCharges['\2'];
			// Calculate charge on N-terminus and add to total charge
			//was charge += 1 / (1 + Math.pow(10, pH - /* 9.53 */8.2));
			// Calculate charge on C-terminus and add to total charge
			//was charge += -1 / (1 + Math.pow(10, 3.2 - pH));

			if (charge > 0.005) {
				lowpH = pH;
			} else if (charge < -0.005) {
				highpH = pH;
			} else {
				break;
			}
			pH = (lowpH + highpH) / 2;
		}
		String s = roundOff(pH, 2);
		htPIcache.put(seq, s);
		return s; // Method returns the pH at which charge is 0 (pI)
	}

	/**
	 * Round off a value to n decimal digits maximum
	 * 
	 * @param mW
	 * @param n
	 * @return
	 */
	private static String roundOff(double mW, int n) {
		String s = "" + mW;
		int pt = s.indexOf('.') + n + 1;
		return (s.length() > pt ? s.substring(0, pt) : s);
	}

}

File: site/swingjs/j2s/main/java/Utilities/Preprocessor.js
(function(){var P$=Clazz.newPackage("main.java.Utilities"),p$1={},I$=[[0,'java.util.Random','java.awt.Color','java.util.Hashtable','java.util.Vector','java.io.File','java.io.PrintWriter','java.io.BufferedWriter','java.io.FileWriter','StringBuffer','main.java.Electro1D.Protein','main.java.Utilities.JSUtil']],I$0=I$[0],$I$=function(i,n){return((i=(I$[i]||(I$[i]=Clazz.load(I$0[i])))),!n&&i.$load$&&Clazz.load(i,2),i)};
/*c*/var C$=Clazz.newClass(P$, "Preprocessor");

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
this.FILE_HEADER="FILE:       ";
this.PROTTITLE_HEADER="NAME:       ";
this.FUNCTION_HEADER="FUNCTION:   ";
this.SEQUENCE_HEADER="SEQUENCE:   ";
this.MOLWT_HEADER="MOLWEIGHT:  ";
this.PIVAL_HEADER="PIVAL:      ";
this.SEQ_SEPARATOR="-----";
},1);

C$.$fields$=[['Z',['is1D','isE2D'],'S',['FILE_HEADER','PROTTITLE_HEADER','FUNCTION_HEADER','SEQUENCE_HEADER','MOLWT_HEADER','PIVAL_HEADER','SEQ_SEPARATOR','inputName','line','e2dOutFileName'],'O',['sequences','java.util.Vector','+sequenceTitles','+piValues','+molecularWeights','+functions','+concentrations','minmax','double[]']]
,['O',['rand','java.util.Random','partialCharges','double[]','+mws','colors','java.awt.Color[]','htPIcache','java.util.Map','+htMWcache']]]

Clazz.newMeth(C$, 'init$',  function () {
}, 1);

Clazz.newMeth(C$, 'randomColor$',  function () {
return C$.colors[C$.rand.nextInt$I(C$.colors.length)];
}, 1);

Clazz.newMeth(C$, 'c$$java_io_File$I',  function (inputFile, fileNum) {
C$.c$$main_java_Electro2D_Electro2D$java_io_File$Z.apply(this, [null, inputFile, fileNum <= 0]);
}, 1);

Clazz.newMeth(C$, 'c$$main_java_Electro2D_Electro2D$java_io_File$Z',  function (e, inputFile, is1D) {
;C$.$init$.apply(this);
this.is1D=is1D;
this.sequences=(e == null  ? Clazz.new_($I$(4,1)) : e.getSequences$());
this.sequenceTitles=(e == null  ? Clazz.new_($I$(4,1)) : e.getSequenceTitles$());
this.molecularWeights=(e == null  ? Clazz.new_($I$(4,1)) : e.getMolecularWeights$());
this.piValues=(is1D ? null : e == null  ? Clazz.new_($I$(4,1)) : e.getPiValues$());
this.functions=(e == null  ? Clazz.new_($I$(4,1)) : e.getFunctions$());
this.inputName=(e == null  ? inputFile.getName$() : e.getLastFileLoaded$());
this.concentrations=(e == null  ? Clazz.new_($I$(4,1)) : null);
}, 1);

Clazz.newMeth(C$, 'writeToE2DFile$S',  function (outFileName) {
if (outFileName == null ) {
outFileName="data" + $I$(5).separator + this.inputName.replace$C$C(".", "_") + ".e2d" ;
}var theFile=Clazz.new_($I$(5,1).c$$S,[outFileName]);
System.out.println$S("Preprocessor.writeToFile creating " + theFile);
if (theFile.exists$()) return;
try {
var out=Clazz.new_([Clazz.new_([Clazz.new_($I$(8,1).c$$java_io_File,[theFile])],$I$(7,1).c$$java_io_Writer), true],$I$(6,1).c$$java_io_Writer$Z);
try {
out.println$S("FILE:       " + this.inputName);
out.println$S("-----");
for (var i=0, n=this.sequences.size$(); i < n; i++) {
out.println$S("NAME:       " + this.sequenceTitles.get$I(i));
C$.writeRecord$java_io_PrintWriter$S$S(out, "FUNCTION:   ", this.functions.get$I(i));
var seq=this.sequences.get$I(i);
C$.writeRecord$java_io_PrintWriter$S$S(out, "SEQUENCE:   ", seq);
out.println$S("MOLWEIGHT:  " + this.molecularWeights.get$I(i));
out.println$S("PIVAL:      " + (this.piValues == null  ? C$.getPI$S(seq) : this.piValues.get$I(i)));
out.println$S("-----");
}

}finally{/*res*/out&&out.close$&&out.close$();}
} catch (e) {
if (Clazz.exceptionOf(e,"java.io.IOException")){
System.err.println$S("Preprocessor.wrteToE2DFile Error writing to file " + e.getMessage$());
return;
} else {
throw e;
}
}
});

Clazz.newMeth(C$, 'writeRecord$java_io_PrintWriter$S$S',  function (out, header, data) {
var length=data.length$();
var pt=0;
while (length > 0){
var n=Math.min(length, 50);
out.println$S(header + data.substring$I$I(pt, pt=pt + n));
length-=n;
}
}, 1);

Clazz.newMeth(C$, 'readFromFile$java_io_BufferedReader$java_util_Vector$java_util_Vector$main_java_Electro2D_Electro2D$I',  function (reader, sequencesOut, proteins, electro2D, fileNum) {
var err=null;
this.isE2D=true;
var buf=Clazz.new_($I$(9,1));
try {
while (true){
if ((this.line=reader.readLine$()).indexOf$S("FILE:       ") != 0) {
err="Preprocessor.readFromeFile Invalid file format.  Missing FILE: line";
break;
}if (electro2D != null ) electro2D.setLastFileLoaded$S(this.line.substring$I(12));
while ((this.line=reader.readLine$()) != null ){
if (this.line.startsWith$S("-")) continue;
this.sequenceTitles.add$O(p$1.readE2D$java_io_BufferedReader$S$StringBuffer.apply(this, [reader, "NAME:       ", buf]));
this.functions.add$O(p$1.readE2D$java_io_BufferedReader$S$StringBuffer.apply(this, [reader, "FUNCTION:   ", buf]));
this.sequences.add$O(p$1.readE2D$java_io_BufferedReader$S$StringBuffer.apply(this, [reader, "SEQUENCE:   ", buf]));
this.molecularWeights.add$O(p$1.readE2D$java_io_BufferedReader$S$StringBuffer.apply(this, [reader, "MOLWEIGHT:  ", buf]));
if (this.is1D) reader.readLine$();
 else this.piValues.add$O(p$1.readE2D$java_io_BufferedReader$S$StringBuffer.apply(this, [reader, "PIVAL:      ", buf]));
}
this.finalizeRead$java_util_Vector$java_util_Vector$main_java_Electro2D_Electro2D$I(sequencesOut, proteins, electro2D, fileNum);
reader.close$();
break;
}
} catch (e) {
if (Clazz.exceptionOf(e,"Exception")){
e.printStackTrace$();
err="Error reading from file.  Be sure you entered the file name correctly. " + e.getMessage$();
} else {
throw e;
}
}
if (err != null ) System.out.println$S(err);
return this.sequences.size$();
});

Clazz.newMeth(C$, 'readE2D$java_io_BufferedReader$S$StringBuffer',  function (reader, key, buf) {
buf.setLength$I(0);
while (this.line.startsWith$S(key)){
buf.append$S(this.line.substring$I(12));
this.line=reader.readLine$();
}
return buf.toString();
}, p$1);

Clazz.newMeth(C$, 'getPIandMW',  function () {
this.minmax=(this.is1D ? null : Clazz.array(Double.TYPE, -1, [-1, 1.7976931348623157E308, -1, 1.7976931348623157E308]));
for (var i=0; i < this.sequences.size$(); i++) {
var mw;
var pI="";
if (this.isE2D) {
mw=this.molecularWeights.get$I(i);
if (!this.is1D) pI=this.piValues.get$I(i);
} else {
var temp=this.sequences.get$I(i);
mw=C$.getMW$S(temp);
this.molecularWeights.add$O(mw);
if (!this.is1D) {
pI=C$.getPI$S(temp);
this.piValues.add$O(pI);
}}if (this.minmax != null ) {
var d=Double.parseDouble$S(mw);
if (d <= this.minmax[0] ) {
this.minmax[0]=d;
}if (d >= this.minmax[1] ) {
this.minmax[1]=d;
}d=Double.parseDouble$S(pI);
if (d <= this.minmax[2] ) {
this.minmax[2]=d;
}if (d >= this.minmax[3] ) {
this.minmax[3]=d;
}}}
}, p$1);

Clazz.newMeth(C$, 'finalizeRead$java_util_Vector$java_util_Vector$main_java_Electro2D_Electro2D$I',  function (sequencesOut, proteins, electro2D, fileNum) {
p$1.getPIandMW.apply(this, []);
var nSeq=this.sequences.size$();
switch (fileNum) {
case -1:
sequencesOut.addAll$java_util_Collection(this.sequences);
return nSeq;
case 0:
for (var i=0; i < nSeq; i++) {
proteins.add$O(Clazz.new_([this.sequenceTitles.get$I(i), "", "", (Double.parseDouble$S(this.molecularWeights.get$I(i))|0), C$.randomColor$()],$I$(10,1).c$$S$S$S$I$java_awt_Color));
if (this.concentrations.size$() > 0) proteins.get$I(i).setConcentration$I(Integer.parseInt$S(this.concentrations.get$I(i)));
}
return nSeq;
case 1:
electro2D.setSequences$java_util_Vector(this.sequences);
electro2D.setSequenceTitles$java_util_Vector(this.sequenceTitles);
electro2D.setMolecularWeights$java_util_Vector(this.molecularWeights);
electro2D.setPiValues$java_util_Vector(this.piValues);
electro2D.setFunctionValues$java_util_Vector(this.functions);
break;
case 2:
electro2D.setSequences2$java_util_Vector(this.sequences);
electro2D.setSequenceTitles2$java_util_Vector(this.sequenceTitles);
electro2D.setMolecularWeights2$java_util_Vector(this.molecularWeights);
electro2D.setPiValues2$java_util_Vector(this.piValues);
electro2D.setFunctionValues2$java_util_Vector(this.functions);
break;
}
electro2D.setLastFileLoaded$S(this.inputName);
electro2D.setMaxAndMinVals$D$D$D$D(this.minmax[1], this.minmax[0], this.minmax[3], this.minmax[2]);
if (!this.isE2D && !$I$(11).isJS ) this.writeToE2DFile$S(this.e2dOutFileName);
return nSeq;
});

Clazz.newMeth(C$, 'getMW$S',  function (seq) {
var wt=C$.htMWcache.get$O(seq);
if (wt != null ) return wt;
var len=seq.length$();
var mW=18;
for (var f=0; f < len; f++) {
mW+=C$.mws[(seq.charCodeAt$I(f))];
}
var mWstring=C$.roundOff$D$I(mW, 2);
C$.htMWcache.put$O$O(seq, mWstring);
return mWstring;
}, 1);

Clazz.newMeth(C$, 'calculatePartialCharges$D',  function (pH) {
C$.partialCharges[68]=C$.getPartialCharge$Z$D$D(false, 4.05, pH);
C$.partialCharges[67]=C$.getPartialCharge$Z$D$D(false, 9.0, pH);
C$.partialCharges[69]=C$.getPartialCharge$Z$D$D(false, 4.75, pH);
C$.partialCharges[89]=C$.getPartialCharge$Z$D$D(false, 10.0, pH);
C$.partialCharges[82]=C$.getPartialCharge$Z$D$D(true, 12.0, pH);
C$.partialCharges[72]=C$.getPartialCharge$Z$D$D(true, 5.98, pH);
C$.partialCharges[75]=C$.getPartialCharge$Z$D$D(true, 10.0, pH);
C$.partialCharges[1]=C$.getPartialCharge$Z$D$D(true, 8.2, pH);
C$.partialCharges[2]=C$.getPartialCharge$Z$D$D(false, 3.2, pH);
}, 1);

Clazz.newMeth(C$, 'getPartialCharge$Z$D$D',  function (isBasic, pK, pH) {
return 1 / (isBasic ? (1 + Math.pow(10, pH - pK)) : -(1 + Math.pow(10, pK - pH)));
}, 1);

Clazz.newMeth(C$, 'getPI$S',  function (seq) {
var pi=C$.htPIcache.get$O(seq);
if (pi != null ) return pi;
var lowpH=0;
var pH=7;
var highpH=14;
var plen=seq.length$();
var charge=0;
while (true){
C$.calculatePartialCharges$D(pH);
charge=0;
for (var a=plen; --a >= 0; ) {
charge+=C$.partialCharges[(seq.charCodeAt$I(a))];
}
charge+=C$.partialCharges[1] + C$.partialCharges[2];
if (charge > 0.005 ) {
lowpH=pH;
} else if (charge < -0.005 ) {
highpH=pH;
} else {
break;
}pH=(lowpH + highpH) / 2;
}
var s=C$.roundOff$D$I(pH, 2);
C$.htPIcache.put$O$O(seq, s);
return s;
}, 1);

Clazz.newMeth(C$, 'roundOff$D$I',  function (mW, n) {
var s="" + new Double(mW).toString();
var pt=s.indexOf$I(".") + n + 1 ;
return (s.length$() > pt ? s.substring$I$I(0, pt) : s);
}, 1);

C$.$static$=function(){C$.$static$=0;
C$.rand=Clazz.new_($I$(1,1));
C$.partialCharges=Clazz.array(Double.TYPE, [256]);
C$.mws=Clazz.array(Double.TYPE, [256]);
{
C$.mws[65]=71.0938;
C$.mws[67]=103.1538;
C$.mws[68]=115.1036;
C$.mws[69]=129.1304;
C$.mws[70]=147.1914;
C$.mws[71]=57.067;
C$.mws[72]=137.156;
C$.mws[73]=113.1742;
C$.mws[75]=128.1888;
C$.mws[76]=113.1742;
C$.mws[77]=131.2074;
C$.mws[78]=114.1188;
C$.mws[80]=97.1316;
C$.mws[81]=128.1456;
C$.mws[82]=156.2022;
C$.mws[83]=87.0932;
C$.mws[84]=101.12;
C$.mws[86]=99.1474;
C$.mws[87]=186.228;
C$.mws[89]=163.1908;
};
C$.colors=Clazz.array($I$(2), -1, [$I$(2).black, $I$(2).blue, $I$(2).cyan, $I$(2).green, $I$(2).magenta, $I$(2).orange, $I$(2).pink, $I$(2).red]);
C$.htPIcache=Clazz.new_($I$(3,1));
C$.htMWcache=Clazz.new_($I$(3,1));
};

Clazz.newMeth(C$);
})();
;Clazz.setTVer('3.3.1-v4');//Created 2022-03-22 08:41:08 Java2ScriptVisitor version 3.3.1-v4 net.sf.j2s.core.jar version 3.3.1-v4
